### PR TITLE
docs: network deployment and release testing

### PR DESCRIPTION
**Pull request by shumkov**
* Added documentation describing network environments and deployment strategies.
* Introduced a comprehensive release testing strategy outlining test types, processes for different release stages, and criteria for promotion between release phases.

## Summary by CodeRabbit

* **Documentation**
  * Added detailed documentation outlining network deployment strategies, including descriptions of Mainnet, Testnet, and Devnet environments.
  * Introduced a comprehensive release testing strategy, covering automated and manual testing stages for Botanix software releases, with criteria for different release types and promotion processes.

### CODE REVIEW

File: adrs/release.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+- how we release
+- semantic versioning
+- networks deployment
+- testing
```
Comment (lamafab) [inline adrs/release.md]: Suggestion; we tag a release for `develop` once a week (or bi-weekly), then we deploy that tagged release on testnet and keep it running there for at least one week. In the meantime we continue working on top of `develop`. After one week, if everything looks fine, we merge that tag into `main`, and we create a new tag for `develop` and deploy that to testnet. Same cycle over again.

Simple as that. But someone has to take ownership of that :)

**Review by lamafab — COMMENTED** [review]
So this does look good to me, although I do wonder whether we're overly strict on the commit message convention? It does make sense for branches and PR's, but maybe this is a little overkill. Up for debate, I'm willing to accept this.

EDIT: Just followed that convention on my last two PR's, it's fine :)

File: adrs/gitflow.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+```
+main
+↑  ↑
+↑  develop
+↑  ↑
+<type>/[scope]/<name>
+```
+
+- `main` - The latest stable version of the code. It must be production ready and can be deployed to production at any time.
+- `develop` - Integration branch for a new release. It contains all the new features and bug fixes. This is the default branch for the repository.
```
Comment (scottmillner) [inline adrs/gitflow.md]: i named the branch `dev`
i can deprecate that branch and rename `develop` if we want

File: adrs/release.md (hunk 1 above)
Comment (scottmillner) [reply adrs/release.md]: sounds good: i can take ownership of that

**Review by scottmillner — DISMISSED** [review]

File: docs/botanix/network-deployment.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+## Networks
+
+### Mainnet
+
+Mainnet is our primary production environment that facilitates real transactions and interactions with the Bitcoin mainnet network.
+It requires the highest level of security and stability.
+
+- Permission-based network (only approved operators can run nodes)
+- Future roadmap includes transition to a permissionless model
+- Infrastructure: 16 federation nodes and 16 RPC nodes
```
Comment (scottmillner) [inline docs/botanix/network-deployment.md]: The number of RPC nodes isn't fixed and depends on external providers.  We have 2 external RPC providers, we run at least one and may need to spin up more to support partners on an as-needed basis.

File: docs/botanix/network-deployment.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+### Testnet
+
+Testnet serves as a testing environment for release candidates that closely mirrors mainnet functionality.
+Business partners, developers, and community members use it to test applications and integrations before deploying to mainnet.
+
+- Considered a production-grade network requiring stability and high availability
+- Will support 3rd party testnet nodes in the future
+- Infrastructure: 15 federation nodes and 15 RPC nodes
```
Comment (scottmillner) [inline docs/botanix/network-deployment.md]: same comment as above in regards to RPC nodes

and current testnet is 3 Federation nodes

File: docs/botanix/release-strategy.md [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+- Contributor acknowledgment
+
+## Release Schedule
+
+We follow a predictable four-week release cycle, composed of two consecutive two-week sprints.
+This schedule balances regular feature delivery with adequate testing time to ensure stability.
+
+```
+Week 1-2                  Week 3-4                  Week 5-6                  Week 7-8
+┌─────────────────────┐   ┌─────────────────────┐   ┌─────────────────────┐   ┌─────────────────────┐
```
Comment (scottmillner) [inline docs/botanix/release-strategy.md]: For Week 3-4, this reads as the entire sprint is for testing?
Whereas I imagine the sprint will involve RC testing and continued new development on `dev` as you mention in `Development Cycle Details`

File: docs/botanix/release-testing.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
+# Release Testing Strategy
+
+## Test Types
+
+### Smoke Tests
```
Comment (scottmillner) [inline docs/botanix/release-testing.md:16]: When I do smoke tests, I always do a pegin and pegout then a simple transfer to ensure basic functionality behaves as expected. We could also require a transaction that interacts with the yield contract:
- https://yield.botanixlabs.dev/

But these tests seem more related to the `Manual Testing` section.

Are you envisioning smoke tests are automated as opposed to manual?

File: docs/botanix/release-testing.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+- Essential services are operational
+
+These lightweight tests identify obvious failures early before investing time in more comprehensive testing.
+The specific implementation details should be detailed and proposed as an additional ADR.
+
+### Automated Test Suite
+
+Our comprehensive automated test suite provides systematic verification of system functionality and consists of two primary components:
+
+**Functional Tests**
```
Comment (scottmillner) [inline docs/botanix/release-testing.md]: When I refer to functional tests, I specifically mean the tests in sidecar:
- https://github.com/botanix-labs/Side-Car/blob/b33ebe30ea33f961c35b1113edc27cfcec0cb1ce/functional-tests/index.ts#L1648

But they are more like E2E tests. 

We can add a test that interacts with the yield contracts as well.

File: docs/botanix/release-testing.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+### Blockchain Compatibility Testing
+
+Compatibility testing ensures that a new release maintains backward compatibility with existing blockchain data and can successfully interact with stable networks.
+
+This testing involves:
+
+1. Syncing a new node from genesis block to current height using the updated software
+2. Verifying that all historical blocks are processed correctly
+3. Ensuring compatibility with both testnet and mainnet blockchains
```
Comment (scottmillner) [inline docs/botanix/release-testing.md]: Until we deprecate Testnet V1 (which has hardforks), we'll have to rely on devnet to test syncing from genesis.

File: docs/botanix/network-deployment.md (hunk 3 above)
Comment (shumkov) [reply docs/botanix/network-deployment.md]: Is would be nice to have at least one ours so we could catch full nodes issues if any happened. What do you think?

File: docs/botanix/network-deployment.md (hunk 4 above)
Comment (shumkov) [reply docs/botanix/network-deployment.md]: Shall we increase the number of nodes to mimic the mainnet.

File: docs/botanix/release-strategy.md (hunk 5 above)
Comment (shumkov) [reply docs/botanix/release-strategy.md]: Yeah, it's a mistake in diagram. I will fix it

File: docs/botanix/release-testing.md (hunk 6 above)
Comment (shumkov) [reply docs/botanix/release-testing.md:16]: Yeah, I was referring to automatic smoke tests. Some quick, simple checks so we see problems at an early stage: component failed, infra problem, port is closed...

File: docs/botanix/release-testing.md (hunk 7 above)
Comment (shumkov) [reply docs/botanix/release-testing.md]: Yeah, something like that. Functional tests could be a smart contract or RPC endpoint testing. The important thing here is to do not duplicate our unit and integration tests. Tests on this level should indicate problems that could happen when we have many nodes and entire system up. I believe VM in nutshell should always behave the same no matter what networks setup is. Otherwise, it should be tested with this test suite as well (we might have some contracts with external params involved)

File: docs/botanix/release-testing.md (hunk 8 above)
Comment (shumkov) [reply docs/botanix/release-testing.md]: Will it be possible to rest the chain (this approach is what we discussed) so we can keep the state we have already instead of deprecation?

File: docs/botanix/network-deployment.md (hunk 4 above)
Comment (scottmillner) [reply docs/botanix/network-deployment.md]: per our offline conversation:
- yes, i think it makes sense to increase the nodes on testnet to match mainnet
- there are several things to consider which @shumkov will explore in a spec/poc

File: docs/botanix/release-testing.md (hunk 8 above)
Comment (scottmillner) [reply docs/botanix/release-testing.md]: per our offline conversation:
- @shumkov will explore truncating the cometbft state so we can sync testnet to identify breaking changes
- this will be done in a spec/poc

**Review by scottmillner — DISMISSED** [review]

File: docs/botanix/release-strategy.md [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+
+## Branches
+
+### main
+- Contains the latest stable version of the code
+- Production-ready and can be deployed at any time
+- Can only receive merges from `hotfix` or `rc` branches
+- Only code owners should be able to merge to `main`
+- Each merge to `main` gets tagged with a version number and triggers a release
+- After merging to `main`, changes must be back-merged to `hotfix`, `develop`, and `rc`
```
Comment (lamafab) [inline docs/botanix/release-strategy.md]: > back-merged

(or rebased)

File: docs/botanix/release-strategy.md [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+
+Release notes should include:
+- Version number and release date
+- Highlights of the most important changes
+- Upgrade instructions if needed
+- Known issues or limitations
+- Link to the full change log
+- Contributor acknowledgment
+
+## Release Schedule
```
Comment (lamafab) [inline docs/botanix/release-strategy.md]: So, I think having a release schedule like this is good, but making plans such as having week 1-2 dedicated to development, and week 3 and 4 dedicated to testing, is not realistic for us. At the current stage, the nature of our work inherently follows a rolling-release schedule.

File: docs/botanix/release-strategy.md [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+- **PATCH**: Incremented for backward-compatible bug fixes
+- **PRERELEASE**: Optional identifier for pre-releases (e.g., `-beta.1`, `-rc.2`, `-hotfix.1`)
+- **BUILD**: Optional build commit SHA (e.g., `+a30ddd1`)
+
+## Branches
+
+### main
+- Contains the latest stable version of the code
+- Production-ready and can be deployed at any time
+- Can only receive merges from `hotfix` or `rc` branches
```
Comment (lamafab) [inline docs/botanix/release-strategy.md]: > Can only receive merges from `hotfix` or `rc` branches

I do wonder; can this somehow be enforced by Github itself?

**Review by lamafab — DISMISSED** [review]
Very well explained! 🚀🚀

File: docs/botanix/release-strategy.md (hunk 11 above)
Comment (shumkov) [reply docs/botanix/release-strategy.md]: Either branch rules or github actions

File: docs/botanix/release-strategy.md (hunk 10 above)
Comment (shumkov) [reply docs/botanix/release-strategy.md]: Please elaborate more on this.

It doesn't mean we stop development in the 3rd and 4th weeks. The idea is that we finalize what we have accomplished and release it. Release what we have ready.

Comment (lamafab) [reply docs/botanix/release-strategy.md]: Yeah, this makes sense :)

File: docs/botanix/network-deployment.md (hunk 3 above)
Comment (birozuru) [reply docs/botanix/network-deployment.md]: We already have our own rpc.

**Review by DariusParvin — COMMENTED** [review]
Really useful reference, @shumkov . Thanks for putting this together! Just a couple questions about the diagram

File: docs/botanix/release-strategy.md [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+  ┌────────────────────────┐   │
+  │        develop         │   │     v1.1.0-alpha.1-a30ddd1, v1.1.0-alpha.2-a30ddd1
+  └───────────┬────────────┘   │
+              │         ▲      │
+              │         │      │
+              ▼         │      │
+  ┌────────────────────────┐   │
+  │       feat/...         ├───┘     A new PR branch
+  └────────────────────────┘
+```
```
Comment (DariusParvin) [inline docs/botanix/release-strategy.md]: Apologies if im misinterpreting the doc but I think the diagram needs to be updated to match the text. Here are my suggestions:

arrows to be added: 
- from main to develop
- from hot fix to main (bidirectional)
- from 'A new PR branch' to hotfix

arrows to be removed: 
- from rc to develop
- from 'A new PR branch' to rc

**Review by rwlockbg — DISMISSED** [review]
Nicely written, I am personally fine with the proposal. I feel that the practice will show to us if it is easy and people will embrace the new flow or we will need to deviate from it.

File: docs/botanix/release-strategy.md (hunk 12 above)
Comment (shumkov) [reply docs/botanix/release-strategy.md]: You are right! Thank you for pointing this out.

Comment (DariusParvin) [reply docs/botanix/release-strategy.md]: updated version looks great! 👍

**Review by SatwikPrabhu — DISMISSED** [review]

File: docs/botanix/release-strategy.md [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+
+## Release Notes
+
+Release notes should include:
+- Version number and release date
+- Highlights of the most important changes
+- Upgrade instructions if needed
+- Known issues or limitations
+- Link to the full change log
+- Contributor acknowledgment
```
Comment (DariusParvin) [inline docs/botanix/release-strategy.md]: I think for external contributors this is nice, and may encourage contributions too. To cover all the team members perhaps we could say 'Botanix inc' or something like that.

File: docs/botanix/release-strategy.md [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+### Release Candidate Release
+1. Merge `develop` branch into `rc` (fast-forward merge)
+2. Bump package versions (e.g., v1.1.0-rc.1)
+3. Update change log
+4. Create release on GitHub from `rc` branch with the `rc` pre-release (e.g., `v1.1.0-rc.1`)
+5. Build artifacts
+6. Push artifacts
+
+### Hotfix Release
+1. Bump package versions (e.g., `v1.0.1-hotfix.1`)
```
Comment (shumkov) [inline docs/botanix/release-strategy.md]: commit sha is missing

File: docs/botanix/release-testing.md [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+- Verify proper interaction between all system components
+- Validate behavior from an external user's perspective
+
+## Sidecar tests
+
+Currently, while we don't have the test suite described above, we have to use our sidecar functional tests that cover some functionality and edge cases.
+
+Sidecar functional tests: https://github.com/botanix-labs/Side-Car/tree/main/functional-tests
+
+TODO: @scoottay instructions to run them.
```
Comment (shumkov) [inline docs/botanix/release-testing.md:54]: @scottmillner instructions to run them.

File: docs/botanix/release-testing.md [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+
```
Comment (shumkov) [inline docs/botanix/release-testing.md:55]: Add performance tests

**Review by lamafab — APPROVED** [review]

**Review by scottmillner — APPROVED** [review]
